src/utils/string.cpp

```diff
@@ -74,6 +75,8 @@ std::string lower(const std::string& str) {
 
 std::string
 get_base_url(const std::string& url) {
+    Logger::init_logging()->log(LogLevel::ERROR, LogOutput::FILE,
```

Comment:
Remove this. This is util. Ideally, utils do not care about logging. Util stuff should return the result (whether it is successful or not), and it is the caller's job to handle these result. So logging should be done by the caller instead.